### PR TITLE
Initial commit k8s_crd CoreDNS plugin

### PR DESCRIPTION
**Pull request by k0da**
Signed-off-by: Dinar Valeev <dinar.valeev@absa.africa>

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,29 @@
+FROM golang:1.14.4 as builder
```
Comment (kuritka) [inline Dockerfile:1]: - It is new project; start  with GO 1.15 (DockerFile, mod files)

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,16 @@
+module github.com/AbsaOSS/k8s_crd
+
+go 1.14
```
Comment (kuritka) [inline go.mod:3]: - It is new project; start  with GO 1.15 (DockerFile, mod files)

File: terratest/test/go.mod [hunk 3]
```diff
@@ -0,0 +1,10 @@
+module testterratest
+
+go 1.14
```
Comment (kuritka) [inline terratest/test/go.mod:3]: - It is new project; start  with GO 1.15 (DockerFile, mod files)

File: terratest/test/go.mod [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+require (
+	github.com/gruntwork-io/terratest v0.26.1
+	github.com/stretchr/testify v1.5.1
+	k8s.io/api v0.17.0
```
Comment (kuritka) [inline terratest/test/go.mod:8]: - terratest .mod file has same modules but different versions than main .mod file. Please unify.

File: terratest/test/helper.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+import (
+	"fmt"
+	"strings"
+	"sort"
+	"testing"
+	"github.com/gruntwork-io/terratest/modules/shell"
+)
+
+func Dig(t *testing.T, dnsServer string, dnsPort int, dnsName string) ([]string, error) {
```
Comment (kuritka) [inline terratest/test/helper.go:11]: just comment: We have very similar function in k8gb. Soon, we should move into AbsaOSS/gopg and reuse

File: terratest/test/basic_test.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+        "github.com/gruntwork-io/terratest/modules/k8s"
+        "github.com/gruntwork-io/terratest/modules/random"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func TestBasicExample(t *testing.T) {
```
Comment (kuritka) [inline terratest/test/basic_test.go:18]: Can we cover by more than one test ? Can we test corner case(s)

**Review by kuritka — CHANGES_REQUESTED** [review]
Consider using linter in Makefile, then `make lint` and lint the project: see https://github.com/AbsaOSS/k8gb/blob/master/Makefile#L234

The code and project structure is very nice.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,70 @@
 # k8s_crd
```
Comment (somaritane) [inline README.md:1]: What `_crd` means in this context, CRD or CoreDNS? 
What do you think about `k8s_dnsendpoint` as the name, to better reflect the usage of plugin?

File: README.md [hunk 8]
```diff
@@ -1,2 +1,70 @@
...
+
+The recommended installation method is using the helm chart provided in the repo:
+
+```
+helm install exdns ./charts/coredns
+```
+## Configure
+
+```
+k8s_gateway [ZONE...]
```
Comment (somaritane) [inline README.md:30]: should it be `k8s_crd`?

File: README.md [hunk 9]
```diff
@@ -1,2 +1,70 @@
...
+## Configure
+
+```
+k8s_gateway [ZONE...] 
+```
+
+Optionally, you can specify what kind of resources to watch, default TTL to return in response and a default name to use for zone apex, e.g.
+
+```
+k8s_gateway example.com {
```
Comment (somaritane) [inline README.md:36]: should it be k8s_crd?

**Review by somaritane — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (k0da) [reply Dockerfile]: done

File: go.mod (hunk 2 above)
Comment (k0da) [reply go.mod]: done

File: terratest/test/go.mod (hunk 3 above)
Comment (k0da) [reply terratest/test/go.mod:3]: done

File: terratest/test/go.mod (hunk 4 above)
Comment (k0da) [reply terratest/test/go.mod]: done

File: README.md (hunk 7 above)
Comment (k0da) [reply README.md:1]: CRD, I had discussion about it with @ytsarev we agreed to name it as CRD. Project had initial name k8s_dnsendpoint (or k8s_extdns)

File: README.md (hunk 8 above)
Comment (k0da) [reply README.md]: fixed

File: README.md (hunk 9 above)
Comment (k0da) [reply README.md]: fixed

File: terratest/test/go.mod (hunk 4 above)
Comment (k0da) [reply terratest/test/go.mod]: have to fix to 0.19.3 as terratest doesn't work with 0.20.2 yet

**Review by kuritka — APPROVED** [review]
would be better to split a large test into smaller tests, but coverage is now there... 

Otherwise LGTM
